README
For an index of all ASC documentation please refer to the file doc/index.html 


## Building on Windows

    cmake --preset windows

ChangeLog
ASC 2.5.0.0

  Major Features:
    - Added function for long distance movement (longer than 1 turn)
    - savegames have preview image

  Minor Features:
    - add: infos about the health and the fuel can be shown at the top of the unit
    - Map transformations can now transform Techadapters
    - unit info shows 3D image of unit
    - more LUA commands
    - Automatic Ammo Production is now on by default for new ASC installations
    - better detection of screen resolution, 100% zoom now default

  Bugfixes:
    - fixed view calcuation problems
    - fixed compilation problems with gcc 4.7
    - Fixed crash on windows when trying to perform invalid resource/ammo transfers through Lua
    - Fixed damage increase through repair
    - Fixed AI crash
    - Fixed positioning of map info window
    - Fixed writing empty service commands
    - Fixed crash when jumping into RF
    - Fixed undo errors when destroying objects
    - and many more...


The old Changelogs were automatically created from CVS history:


2009-12-20 17:34  mbickel

	* configure, configure.in, data/ascdatacoreversion.asctxt,
	  installer/asc.nsi, source/strtmesg.cpp:  Changed version number
	  to 2.4

2009-12-20 17:27  mbickel

	* data/campaign/tutorial/: tut000.map.fr_FR, tut004.map.fr_FR,
	  tut007.map.fr_FR, tut008.map.fr_FR, tut010.map.fr_FR,
	  tut013.map.fr_FR, tut014.map.fr_FR, tut017.map.fr_FR,
	  tut018.map.fr_FR, tut019.map.fr_FR, tut023.map.fr_FR:  Fixed
	  syntax error in franch tutorial translation

2009-12-19 19:23  mbickel

	* data/campaign/standard/asc009.map:  Added some pipelines for
	  easier connection of conquered depots to net

2009-12-19 17:25  mbickel

	* data/campaign/kamor/: kam001.map, kam002.map, kam003.map,
	  kam004.map, kam005.map, kam006.map, kam007.map, kam008.map,
	  kam009.map, kam010.map, kam011.map, kam012.map, kam013.map,
	  kam014.map, kam015.map, kam016.map, kam017.map, kam018.map,
	  kam019.map, kam020.map, kam021.map, kam022.map, kam023.map,
	  kam024.map, kam025.map:  Updated maps for Kamor campaign

2009-12-19 02:37  grom

	* data/campaign/: nos/nos001.map.de_DE, nos/nos001.map.fr_FR,
	  nos/nos002.map.fr_FR, nos/nos003.map.fr_FR, nos/nos004.map.fr_FR,
	  nos/nos005.map.fr_FR, nos/nos006.map.fr_FR, nos/nos007.map.fr_FR,
	  nos/nos008.map.fr_FR, standard/asc003.map.de_DE,
	  standard/asc014.map.de_DE, tutorial/tut000.map.fr_FR,
	  tutorial/tut001.map.fr_FR, tutorial/tut002.map.fr_FR,
	  tutorial/tut003.map.fr_FR, tutorial/tut004.map.fr_FR,
	  tutorial/tut005.map.fr_FR, tutorial/tut006.map.fr_FR,
	  tutorial/tut007.map.fr_FR, tutorial/tut008.map.fr_FR,
	  tutorial/tut009.map.fr_FR, tutorial/tut010.map.fr_FR,
	  tutorial/tut011.map.fr_FR, tutorial/tut012.map.fr_FR,
	  tutorial/tut013.map.fr_FR, tutorial/tut014.map.fr_FR,
	  tutorial/tut015.map.fr_FR, tutorial/tut016.map.fr_FR,
	  tutorial/tut017.map.fr_FR, tutorial/tut018.map.fr_FR,
	  tutorial/tut019.map.fr_FR, tutorial/tut020.map.fr_FR,
	  tutorial/tut021.map.fr_FR, tutorial/tut022.map.fr_FR,
	  tutorial/tut023.map.fr_FR, tutorial/tut024.map.fr_FR,
	  tutorial/tut025.map.fr_FR: translations

2009-12-19 01:56  grom

	* data/campaign/tutorial/tut030.map.en_EN: translation

2009-12-18 22:46  mbickel

	* data/campaign/tutorial/tut024.map:  Added some mineral resources
	  to discover

2009-12-18 19:53  mbickel

	* data/: buildings/asc/bld_asc_atomkraftwerk.asctxt,
	  campaign/tutorial/tutorial.asccampaign:  Rename nuclear power
	  plant
	   Change tut campaign description

2009-12-18 14:25  mbickel

	* data/campaign/tutorial/: tut006.map.en_EN, tut009.map.en_EN,
	  tut030.map, tut030.map.en_EN:  Added translations
	   Removed the SAM-Turret part of tut030 because it doesn't work
	  any more

2009-12-17 23:58  mbickel

	* data/Makefile.am, data/Makefile.in, source/packagemanager.cpp,
	  source/packagerepository.cpp, source/packagerepository.h:  Fixed
	  build system
	   Better error message when packages not found

2009-12-17 22:29  mbickel

	* data/campaign/tutorial/tut019.map.en_EN:  More translation

2009-12-17 20:33  mbickel

	* configure, configure.in, data/Makefile.am, data/Makefile.in,
	  installer/asc.nsi, source/loaders.cpp, source/strtmesg.cpp:
	  Updated version number and Makefiles

2009-12-17 19:08  mbickel

	* data/campaign/tutorial/: tut000.map.en_EN, tut021.map.en_EN,
	  tut022.map.de_DE, tut022.map.en_EN, tut023.map.en_EN,
	  tut024-mining-bars.png, tut024.map.de_DE, tut024.map.en_EN,
	  tut025.map:  More translations

2009-12-16 22:56  mbickel

	* data/campaign/tutorial/tut025.map.en_EN:  Translation

2009-12-16 22:40  mbickel

	* data/campaign/tutorial/: tut010.map.en_EN, tut016.map.en_EN,
	  tut018.map.en_EN, tut020.map.en_EN:  Still more translations

2009-12-16 21:26  mbickel

	* data/campaign/tutorial/: tut018.map.de_DE, tut018.map.en_EN:
	  Even more translation
	   Fix the description of recycling

2009-12-16 19:52  mbickel

	* data/campaign/tutorial/tut016.map.en_EN:  Translation

2009-12-16 19:37  mbickel

	* data/campaign/tutorial/tut011.map.en_EN:  Translation

2009-12-16 18:59  mbickel

	* data/campaign/tutorial/tut015.map.en_EN:  Another translation

2009-12-16 18:39  mbickel

	* data/campaign/tutorial/: tut013.map.en_EN, tut014.map.en_EN:
	  Translations again

2009-12-16 18:31  mbickel

	* data/campaign/tutorial/: tut012.map.en_EN, tut013.map.en_EN:
	  More tutorial translations

2009-12-16 17:36  mbickel

	* data/campaign/tutorial/: tut017.map.de_DE, tut017.map.en_EN:
	  More translations

2009-12-16 12:50  mbickel

	* data/campaign/tutorial/: tut000.map.en_EN, tut001.map.en_EN,
	  tut005.map.en_EN, tut007.map.en_EN, tut008.map.en_EN,
	  tut010.map.en_EN:  Added and improved english tutorial
	  translation

2009-12-15 22:01  mbickel

	* Makefile.in, asc.kdevelop, configure, configure.in,
	  doc/Makefile.in, doc/manpages/Makefile.in,
	  doc/unitguide/Makefile.in, doc/unitguide/output/Makefile.in,
	  installer/Makefile.in, source/Makefile.in,
	  source/guifunctions.cpp, source/actions/Makefile.in,

source/pd.h
/*! \file pd.h
    \brief The interface of the Pulldown Menu
*/


//     $Id: pd.h,v 1.11.2.1 2006-01-22 18:56:36 mbickel Exp $
//
//     $Log: not supported by cvs2svn $
//     Revision 1.12  2006/01/22 10:33:50  mbickel
//      Fixed compilation problems with SUSE
//
//     Revision 1.11  2004/07/12 18:15:08  mbickel
//      Lots of tweaks and fixed for more portability and compilation with mingw
//
//     Revision 1.10  2002/01/19 22:39:19  mbickel
//      Added music playing functions to ASC
//
//     Revision 1.9  2002/01/07 11:40:40  mbickel
//      Fixed some syntax errors
//
//     Revision 1.8  2001/07/30 17:43:13  mbickel
//      Added Microsoft Visual Studio .net project files
//      Fixed some warnings
//
//     Revision 1.7  2001/01/28 14:04:15  mbickel
//      Some restructuring, documentation and cleanup
//      The resource network functions are now it their own files, the dashboard
//       as well
//      Updated the TODO list
//
//     Revision 1.6  2000/08/12 12:52:50  mbickel
//      Made DOS-Version compile and run again.
//
//     Revision 1.5  2000/06/09 10:50:59  mbickel
//      Repaired keyboard control of pulldown menu
//      Fixed compile errors at fieldlist with gcc
//
//     Revision 1.4  2000/05/23 20:40:48  mbickel
//      Removed boolean type
//
//     Revision 1.3  1999/12/28 21:03:18  mbickel
//      Continued Linux port
//      Added KDevelop project files
//
//     Revision 1.2  1999/11/16 03:42:22  tmwilson
//     	Added CVS keywords to most of the files.
//     	Started porting the code to Linux (ifdef'ing the DOS specific stuff)
//     	Wrote replacement routines for kbhit/getch for Linux
//     	Cleaned up parts of the code that gcc barfed on (char vs unsigned char)
//     	Added autoconf/automake capabilities
//     	Added files used by 'automake --gnu'
//
//
/*
    This file is part of Advanced Strategic Command; http://www.asc-hq.de
    Copyright (C) 1994-1999  Martin Bickel  and  Marc Schellenberger

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; see the file COPYING. If not, write to the 
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, 
    Boston, MA  02111-1307  USA
*/


   #define rz 21
   #define rz2 245
   #define maxnamelength 30
   #define textdifference 20
   #define pdfieldenlargement 7

   typedef struct tpdbutton {
          char         name[maxnamelength];
          tkey         shortkey;
          char      tswitch;
          int          actionid;
       } tpdbutton;

   typedef struct tpdfield {
           char            name[maxnamelength];
           int             count;
           dynamic_array<tpdbutton>  button;
           tkey            shortkey;
           int             rtextstart,xstart,xwidth,height;
        } tpdfield;
   typedef tpdfield *ppdfield;

   typedef struct pdbar {
           int     pdbreite;
           dynamic_array<tpdfield>   field;
           int     count;
   } pdbar;

   class tpulldown {
         int pdfieldnr, buttonnr;
      protected:
         bool alwaysOpen;
      public :
         bool barstatus;
         int  action,mousestat;
         int  bkgcolor,textcolor,shortkeycolor,rcolor1,rcolor2;
         int pdfieldtextdistance,textstart,righttextdifference,anf,ende;
         void *backgrnd,*barbackgrnd;
         char lt[100];
         char rt[100];
         pdbar pdb;
         tkey key;
         int action2execute;

         virtual void init(void);
         virtual void setvars(void);
         virtual void baron(void);
         virtual void baroff(void);
         void redraw ();
         virtual void openpdfield(void);
         virtual void closepdfield(void);
         virtual void showbutton(void);
         virtual void hidebutton(void);
         virtual void checkpulldown(void);
         virtual void run(void);
         virtual void setshortkeys(void);
         virtual void lines(int x1,int y1,int x2,int y2);
         virtual void nolines(int x1,int y1,int x2,int y2);
         virtual void getleftrighttext(char *qtext, char *ltext, char *rtext);
         virtual int getpdfieldheight(int pdfieldnr, int pos);
         virtual void done(void);
         void addbutton ( char* name, int id );
         void addfield ( char* name );
         tpulldown ( void );
         virtual ~tpulldown() {};
   };




   extern pfont        pulldownfont;
